### PR TITLE
[SPARK-19151][SQL]DataFrameWriter.saveAsTable support hive overwrite

### PR DESCRIPTION
**Pull request by windpiger**
## What changes were proposed in this pull request?

After [SPARK-19107](https://issues.apache.org/jira/browse/SPARK-19107), we now can treat hive as a data source and create hive tables with DataFrameWriter and Catalog. However, the support is not completed, there are still some cases we do not support.

This PR implement:
DataFrameWriter.saveAsTable work with hive format with overwrite mode

## How was this patch tested?
unit test added

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 1]
```diff
@@ -383,14 +383,25 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
         // Get all input data source relations of the query.
         val srcRelations = df.logicalPlan.collect {
           case LogicalRelation(src: BaseRelation, _, _) => src
+          case relation: CatalogRelation if DDLUtils.isHiveTable(relation.catalogTable) => relation
         }
+
         EliminateSubqueryAliases(catalog.lookupRelation(tableIdentWithDB)) match {
           // Only do the check if the table is a data source table (the relation is a BaseRelation).
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: update this comment?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 2]
```diff
@@ -383,14 +383,25 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
...
-          // TODO(cloud-fan): also check hive table relation here when we support overwrite mode
           // for creating hive tables.
           case LogicalRelation(dest: BaseRelation, _, _) if srcRelations.contains(dest) =>
             throw new AnalysisException(
               s"Cannot overwrite table $tableName that is also being read from")
+          // check hive table relation when overwrite mode
+          case relation: CatalogRelation if DDLUtils.isHiveTable(relation.catalogTable)
+            && srcRelations.filter(_.isInstanceOf[CatalogRelation]).exists {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: doesn't `srcRelations.contains(dest)` work?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala [hunk 3]
```diff
@@ -112,9 +112,9 @@ private[hive] trait HiveStrategies {
         // Currently we will never hit this branch, as SQL string API can only use `Ignore` or
         // `ErrorIfExists` mode, and `DataFrameWriter.saveAsTable` doesn't support hive serde
         // tables yet.
-        if (mode == SaveMode.Append || mode == SaveMode.Overwrite) {
+        if (mode == SaveMode.Append) {
           throw new AnalysisException(
-            "CTAS for hive serde tables does not support append or overwrite semantics.")
+            "CTAS for hive serde tables does not support overwrite semantics.")
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala]: does not support append

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 4]
```diff
@@ -1314,6 +1314,10 @@ class HiveDDLSuite
         .write.format("hive").option("fileFormat", "avro").saveAsTable("t")
       checkAnswer(spark.table("t"), Row(1, "a"))
 
+      Seq(9 -> "x").toDF("i", "j")
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: let's test with a different schema, e.g. `Seq("a" -> 1)...`

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 2 above)
Comment (windpiger) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/MetastoreRelation.scala#L54

MetastoreRelation equals contains output, while srcRelations and catalog.lookupRelation(tableIdentWithDB) have different output.
such as: (a#81,b#82)  <-> (a#85,b#86)

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala (hunk 3 above)
Comment (windpiger) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala]: sorry, fixed it

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: can we update `MetastoreRelation.equals`? Looks like we should not include the `output` in equality check.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/MetastoreRelation.scala [hunk 5]
```diff
@@ -49,8 +49,7 @@ private[hive] case class MetastoreRelation(
   override def equals(other: Any): Boolean = other match {
     case relation: MetastoreRelation =>
       databaseName == relation.databaseName &&
-        tableName == relation.tableName &&
-        output == relation.output
+        tableName == relation.tableName
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/MetastoreRelation.scala]: When we change `equals`, we also need to change `hashCode`. 

`LogicalRelation`'s [`equals`](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala#L56) also checks the output. 

When the underlying Hive tables change the schema, are we able to detect it?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 6]
```diff
@@ -383,14 +383,19 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
         // Get all input data source relations of the query.
         val srcRelations = df.logicalPlan.collect {
           case LogicalRelation(src: BaseRelation, _, _) => src
+          case relation: CatalogRelation if DDLUtils.isHiveTable(relation.catalogTable) => relation
         }
         EliminateSubqueryAliases(catalog.lookupRelation(tableIdentWithDB)) match {
           // Only do the check if the table is a data source table (the relation is a BaseRelation).
-          // TODO(cloud-fan): also check hive table relation here when we support overwrite mode
           // for creating hive tables.
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: This line should also be removed.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 7]
```diff
@@ -383,14 +383,19 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
...
           case LogicalRelation(dest: BaseRelation, _, _) if srcRelations.contains(dest) =>
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: @cloud-fan are we facing the same issue here? We also rely on the default `equals` function here. We did not override the `equal` function for `HadoopFsRelation`

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/MetastoreRelation.scala (hunk 5 above)
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/MetastoreRelation.scala]: if the underlying Hive tables change the schema, I think we will break many places, as we cache the table relation. In the overwrite case, although the table schema changed, they are still the same table.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 7 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: we will cache the table relation for data source table, so this is fine here.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 8]
```diff
@@ -1314,6 +1314,14 @@ class HiveDDLSuite
         .write.format("hive").option("fileFormat", "avro").saveAsTable("t")
       checkAnswer(spark.table("t"), Row(1, "a"))
 
+      Seq("c" -> 1).toDF("i", "j")
+        .write.format("hive").mode(SaveMode.Overwrite).option("fileFormat", "avro").saveAsTable("t")
+      checkAnswer(spark.table("t"), Row("c", 1))
+
+      Seq(9 -> "x").toDF("i", "j")
+        .write.format("hive").mode(SaveMode.Overwrite).option("fileFormat", "avro").saveAsTable("t")
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: try with a different fileFormat here.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 9]
```diff
@@ -1340,8 +1348,8 @@ class HiveDDLSuite
       val e3 = intercept[AnalysisException] {
         spark.table("t").write.format("hive").mode("overwrite").saveAsTable("t")
       }
-      assert(e3.message.contains(
-        "CTAS for hive serde tables does not support append or overwrite semantics"))
+      assert(e3.message.contains("Cannot overwrite table")
+        && e3.message.contains("that is also being read from"))
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: nit: we can test the full message `Cannot overwrite table default.t that is also being read from`

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/MetastoreRelation.scala (hunk 5 above)
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/MetastoreRelation.scala]: Logically, yes.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala (hunk 7 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: We only check `BaseRelation`. Let me check whether we can get rid of the comparison of `output` from the `equals` of `LogicalRelation` and see whether we can hit test case failure.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: eventually we will unify the table relation of data source and hive tables, this problem will go away at that time :)

**Comment by cloud-fan** [comment]
LGTM except https://github.com/apache/spark/pull/16549/files#r95670788

**Comment by windpiger** [comment]
thanks, I will keep track of this after hive and datasource unified .
https://github.com/apache/spark/pull/16549/files#r95670788

**Comment by gatorsmile** [comment]
: ) It could be caused by the latest merge of https://github.com/apache/spark/pull/16233

**Comment by gatorsmile** [comment]
Let us know if you are unable to figure it out. 

BTW, you should update the `hashCode` function too, like what you did in `equals`.

**Comment by windpiger** [comment]
ok  thanks~ I am working to find out the reason why it failed

**Comment by windpiger** [comment]
For test case 
```
  test("intersect") {
    checkSQL("SELECT * FROM t0 INTERSECT SELECT * FROM t0", "intersect")
  }
```

https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/QueryPlanner.scala#L83


when `placeholder` & `childPlan` are `PlanLater MetaStoreRelation[#87] default t0`,
and `candicate p` is `PlanLater MetaStoreRelation[#88] default t0`, 
these two will be equal `without` output compare, that is `p == placeholder`, it will transform to `childPlan`.

That will cause Exception 
`Could't find id#88L in [id#87L]`.

If we don't change the equal, it will be OK.

I think we should not change the MetaStoreRelation equal logic

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 10]
```diff
@@ -383,14 +383,19 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
         // Get all input data source relations of the query.
         val srcRelations = df.logicalPlan.collect {
           case LogicalRelation(src: BaseRelation, _, _) => src
+          case relation: CatalogRelation if DDLUtils.isHiveTable(relation.catalogTable) => relation
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: how about we only collect table identifiers?
```
val srcTables = df.logicalPlan.collect {
  case relation: CatalogRelation if DDLUtils.isHiveTable(relation.catalogTable) => relation.catalogTable.identifier
}
...
  case relation: CatalogRelation if DDLUtils.isHiveTable(relation.catalogTable) &&
      srcTables.contains(relation.catalogTable.identidier) => throw ...
```

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 11]
```diff
@@ -383,14 +383,20 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
         // Get all input data source relations of the query.
         val srcRelations = df.logicalPlan.collect {
           case LogicalRelation(src: BaseRelation, _, _) => src
+          case relation: CatalogRelation if DDLUtils.isHiveTable(relation.catalogTable) =>
+            relation.catalogTable.qualifiedName
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: why not use `identifier`? I think it's more concise to represent table identifier.

**Comment by windpiger** [comment]
ok, changed it,thanks!

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala [hunk 12]
```diff
@@ -112,9 +112,9 @@ private[hive] trait HiveStrategies {
         // Currently we will never hit this branch, as SQL string API can only use `Ignore` or
         // `ErrorIfExists` mode, and `DataFrameWriter.saveAsTable` doesn't support hive serde
         // tables yet.
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala]: Could you also update the comment?

**Comment by gatorsmile** [comment]
LGTM except a minor comment. Thanks!

**Comment by windpiger** [comment]
comment updated

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 13]
```diff
@@ -383,14 +383,20 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
         // Get all input data source relations of the query.
         val srcRelations = df.logicalPlan.collect {
           case LogicalRelation(src: BaseRelation, _, _) => src
+          case relation: CatalogRelation if DDLUtils.isHiveTable(relation.catalogTable) =>
+            relation.catalogTable.identifier
         }
         EliminateSubqueryAliases(catalog.lookupRelation(tableIdentWithDB)) match {
           // Only do the check if the table is a data source table (the relation is a BaseRelation).
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: this comment also need to be updated, we do not "only" check data source tables.

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Pull request closed without merging**